torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -313,7 +313,16 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
       if (profiled_type == TensorType::get()) {
         continue;
       }
-      it->input()->setType(it->ty(attr::profiled_type));
+      // If we encounter non-identical profiled types for the same value, merge
+      // them.  This situation can happen if, e.g., loop unrolling duplicates
+      // profiled types in a loop body in a manner that isn't logically
+      // consistent (see TestTEFuser.test_unrolled_cat).
+      auto input_type = it->input()->type()->expect<TensorType>();
+      if (input_type == TensorType::get()) {
+        it->input()->setType(profiled_type);
+      } else {
+        it->input()->setType(input_type->merge(*profiled_type));
+      }
```

Comment:
I still think the fix is invalid: e.g. this check assumes there would be conflicting types in the graph - but in the test all the types in the graph are either "Float(1)" or "Tensor" which means that we're either disabling the fuser entirely or not catching anything. I believe at this point it is too late(*) to try to find inaccuracies in profiling data - it should be done when that data is introduced and that happens in two places currently: 1) during profiling itself, 2) in loop unroller when it copies profiling data for the subsequent iterations. I'm sure (1) handles conflicting types correctly, but (2) can't do that and hence I think it should just drop the data.

\* _We still can try finding inconsistencies in the profiling data if we employ an alternative mechanism to obtain that info - e.g. shape inference. But without an alternative there is simply nothing to verify against._

Reply:
If you insert `assert(false)` into the else branch, will it ever fire?